posts/2024-08-26-webassembly-targets-and-new-on-by-default-features.md

```diff
@@ -0,0 +1,191 @@
+support the reference-types proposal. It is expected that this change will have
+a low impact due to the age of the reference-types proposal and breadth of
+implementation in engines. Given the multitude of WebAssembly engines, however,
+it's recommended that any WebAssembly users test out Nightly Rust and see if
+the produced module still runs on the engine of choice.
+
+### LLVM, Rust, and Multiple Tables
+
+One interesting point worth mentioning is that despite reference-types enabling
+multiple tables in WebAssembly modules this is not actually taken advantage of
+at this time by either LLVM or Rust. WebAssembly modules emitted will still have
+at most one table of functions. This means that the over-long 5-byte encoding of
+index 0 as `0x80 0x80 0x80 0x80 0x00` is not actually necessary at this time.
+LLD, LLVM's linker for WebAssembly, wants to process all [LEB] relocations in a
+similar manner which currently forces this 5-byte encoding of zero. For example
+when a function calls another function the `call` instruction encodes the target
+function index as a 5-byte [LEB] which is filled in by the linker. There is
+quite often more than one function so the 5-byte encoding enables all possible
+function indices to be encoded.
+
+In the future LLVM might start using multiple tables as well. For example LLVM
+may have a mode in the future where there's a table-per-function type instead of
+a single heterogenous table. This can enable engines to implement
+`call_indirect` more efficiently. This is not implemented at this time, however.
+
+For users who want a minimally-sized WebAssembly module (e.g. if you're in a web
+context and sending bytes over the wire) it's recommended to use an optimization
+tool such as [`wasm-opt`] to shrink the size of the output of LLVM. Even before
+this change with reference-types it's recommended to do this as [`wasm-opt`] can
+typically optimize LLVM's default output even further. When optimizing a module
+through [`wasm-opt`] these 5-byte encodings of index 0 are all shrunk to a
+single byte.
+
+## Enabling Multi-Value by Default
+
+The second feature enabled by default in LLVM 19 is multi-value. The
+[multi-value proposal to WebAssembly][multi-value] enables functions to have
+more than one return value for example. WebAssembly instructions are
+additionally allowed to have more than one return value as well. This proposal
+is one of the first to get merged into the WebAssembly specification after the
+original MVP and has been implemented in many engines for quite some time.
+
+The consequences of enabling this feature by default in LLVM are more minor for
+Rust, however, than enabling reference-types by default. LLVM's default ABI for
+WebAssembly code is not changing even when multi-value is enabled. Additionally
+Rust's ABI is not changing either and continues to match LLVM's. Despite this
+though the change has the possibility of still affecting Nightly users of Rust.
```

Comment:
Suggestion: I think we want to be careful with implied ABI stability:

- It happens that LLVM's WebAssembly ABI was not changed after multi-value proposal was enabled.
- It happens that Rust's WebAssembly ABI did not change either. (But I think we explicitly don't promise ABI stability for `repr(Rust)`)

*But* we probably want a remark about the stability of Rust's WebAssembly ABIs being subject to change, as I can't imagine this can be perma-frozen with new proposals being added and enabled-by-default?

Reply:
I've added a small extra section about this to help clarify in addition to trying to explicitly talk about `extern "C"` is the thing not changing. (but please take a look once I push to review as well)